### PR TITLE
Some refcounting fixes.

### PR DESCRIPTION
**Pull request by haberman**
Clearly this stuff is too complex overall.
The plan is to move away from this and more towards
pools, like proto2 uses.

### CODE REVIEW

**Comment by haberman** [comment]
Review to @scwhittle. cc @teboring.

File: upb/descriptor/reader.c [hunk 1]
```diff
@@ -651,6 +651,7 @@ static void *msg_startext(void *closure, const void *hd) {
   return r;
 }
 
+#include <stdio.h>
```
Comment (scwhittle) [inline upb/descriptor/reader.c]: rm?

File: upb/refcounted.h [hunk 2]
```diff
@@ -28,8 +28,6 @@
  * For this reason we don't enable it by default, even in debug builds.
  */
 
-/* #define UPB_DEBUG_REFS */
```
Comment (scwhittle) [inline upb/refcounted.h]: Can you update preceding comments to name this variable then?

File: upb/descriptor/reader.c [hunk 3]
```diff
@@ -663,9 +664,13 @@ static void *msg_startfield(void *closure, const void *hd) {
 static bool msg_endfield(void *closure, const void *hd) {
   upb_descreader *r = closure;
   upb_msgdef *m = upb_descreader_top(r);
+  bool ok;
   UPB_UNUSED(hd);
 
-  upb_msgdef_addfield(m, r->f, &r->f, NULL);
+  if (upb_fielddef_containingoneof(r->f) == NULL) {
```
Comment (scwhittle) [inline upb/descriptor/reader.c]: add a comment indicating that oneofs fields are added through oneofs table

**Comment by scwhittle** [comment]
back to you

File: upb/descriptor/reader.c (hunk 1 above)
Comment (haberman) [inline upb/descriptor/reader.c]: Done.

File: upb/refcounted.h (hunk 2 above)
Comment (haberman) [inline upb/refcounted.h]: I added it back.

File: upb/descriptor/reader.c (hunk 3 above)
Comment (haberman) [inline upb/descriptor/reader.c]: Done.

**Comment by haberman** [comment]
Thanks! Back to you.

**Comment by scwhittle** [comment]
LGTM

**Pull request merged**